### DISCUSSION TITLE
Support for the 'MayaReference' schema

### DISCUSSION THREAD

**Discussion by pierre-ed** [Autodesk/maya-usd#983] [Q&A] [answered]
Hi everyone!

I have yet another question (which is likely to come from the fact that I easily get confused with the organization of the maya_usd codebase). :)

I wanted to know whether the 'adsk' plugin enables the support of `MayaReference` prims as the 'AL' plugin does.

At first, I thought the `MayaReference` schema was a feature reserved to the 'AL' side of maya_usd, but after seeing mentions of those schemas in files from the `lib/` common directory (most notably, the `UsdMayaTranslatorMayaReference` translator and the aliasing between `ALMayaReference` and `MayaReference`), I began hoping that they were also supported by the 'adsk' plugin. 

Still, after a few trials, I am unable to properly load a Maya reference via such prims through a `mayaUsdProxyShape` node. When imported through a `AL_usdmaya_ProxyShape` node however, the Maya reference is added to the scene as expected and can be confined to a specific variant without issue.

So, are `MayaReference` prims supported by the 'adsk' plugin ? or am I misusing them somehow?
Thank you in advance for your insight!
Cheers,
-- Pierre.

**Comment by kxl-adsk** [comment, marked as answer]
Hi Pierre,

Our intentions are to support `MayaReference` variant switching workflow (like the one you can find in AL plugin). We started by bringing the schema to the core library. This way it can be shared by all studio plugins. We also added an import translator for the schema, so using `adsk` or `pxr` plugins, you can import assets with `MayaReference` schema. The remaining part is related to handling variant switching - this is currently missing in `adsk` plugin.

Cheers,
KxL

**Reply by daviddeacon** [reply]
This was answered quite a while ago, and I was wondering if exposing the MayaReference schema/workflow is being actively developed? As it's something that we would be very interested in adding to our pipeline.

**Reply by kxl-adsk** [reply]
This workflow is definitely on our radar but we can't share any more details at this point. Do you happen to have a list of requirements that you would like us to meet? Anything that you like or don't like in implementation from the AL plugin?

**Reply by daviddeacon** [reply]
I haven't had the opportunity to use the AL plugin, I don't have the technical wizardry to build it and all of its dependancies. I did see a demo of it being used in an online presentation but can't for the life of me find it again.

I'm currently trying to roll my own workflow by using the python API and custom meta data on assets. So USD models have a path to their rig stored in their meta data so an artist can select part of the stage, run a script that reference the Maya rig and deactivates source prim in the stage. The script setups any necessary layering / composition archs based on the stage. They then do their animation / edits, and commit changes via a script that rights back to the USD stage. Well at least that's the theory. The main aim is to have artist/animators not having to do any manual setting up of sublayers or referencing rigs.

It's early but I'm starting to make progress, but finding what I need in the python API has been challenging.

**Comment by pierre-ed** [comment]
Hi!
Thank you very much for your quick answer.

This is great news: glad to know that the `MayaReference` schema is fully supported by the entirety of the maya_usd implementation. The bad news is that I am definitely misusing it then. ;)

Indeed, even on [the most basic of examples](https://github.com/Autodesk/maya-usd/files/5664823/QandA_mayaReference.zip), I cannot trigger the Maya referencing process through `mayaUsdProxyShape` nodes. When I import a single USD layer that directly defines (without any variant switch) a `MayaReference` prim via the 'Create' menu, the prim hierarchy does appear in the outliner but the reference editor remains empty.
This behavior differs from what happens when the very same USD layer is brought in via the 'AL' importing procedure ('USD' menu > 'Proxy shape' > 'Import'): in this case, the Maya reference is correctly handled and its content directly appears in the viewport.
*(I am relying on the 0.5.0 version of maya_usd and working with Maya-2020.1 on a Linux workstation.)*

Am I missing painfully obvious? 
Thanks again for your help.
-- Pierre

**Reply by kxl-adsk** [reply]
MayaReference is supported when importing usd to Maya, i.e. when using `File->Import`. It is not supported yet in native workflows, i.e. when using `Create` menu. This is what I meant by saying the actual variant switching is currently missing in `adsk` plugin and only available in `al` plugin.

**Comment by pierre-ed** [comment]
Hi Krystian.

Sorry for my misunderstanding, you're perfectly right.
The Maya reference does show up through a `File->Import`, but not via a proxyShape. 
I got it. :) Thank you for your help.

Cheers.

**Comment by scott-wilson** [comment]
Hey, I was hoping to get an update on this. It looks like as of version 0.11, trying to load in a referenced Maya scene through the `MayaReference` or `ALMayaReference` doesn't work (sub-scene is not referenced) if you do the create > USD, but file > import/reference works.

Is this improved in a later version, or is that still on the to-do list? If it is in the to-do list, then what's the current recommendation to be able to have a Maya scene referenced from a USD file?

**Reply by kxl-adsk** [reply]
The ability to load MayaReference from a proxy shape (create > USD) is something we are actively working on. 

And just to make sure we are able to validate your use case as well - what is the workflow you have in mind that requires referencing ma/mb files from USD?

**Reply by scott-wilson** [reply]
Right now, we're still very early in the process of adopting USD in our pipeline, so we're still figuring a lot out. But, the main thing we'd like from this is to be able to build sets.

So, assets will create a USD layer in Maya, add in other USD elements for non-deforming geometry (for example, chairs and tables), and reference in Maya scenes for deforming geometry (for example, curtains). The resulting USD scene gets published out.

Animation then loads the set USD and animates the scene, and then we do a USD cache of the whole set.

**Reply by ralphmeijerambassadors** [reply]
Hi,

To prevent duplicate posts I'll just add a response here. We are just doing some research on how to implement usd in our pipeline and being able to load a mayaReference in a proxy shape is something we would really like to implement. I noticed as of version 0.16.0 there should be the ability to add a mayaReference to prim. In the maya UI I'm able to do this and it create a mayaReference node under the prim. However I can't select a maya file to actually load in the reference in the UI. I also noticed a script called mayaUsdAddMayaReference.py in maya-usd/lib/mayaUsd/resources/scripts/ with a function that seems to create a mayaReference under a specified prim. How would I go about the use this function. I've tried digging through all the functions in the API but I can't seem to find anything. 

I'm sorry if I'm asking something very obvious. I'm very new to this whole workflow and API.

Thanks in advance!

Ralph

**Reply by ralphmeijerambassadors** [reply]
Sorry I think I misunderstood the overall workflow and concepts of using the mayaReferencePrim. After more digging it clicked what the actual workflow is. Still need to work things out a bit more but it seems to work. 

Thanks

**Reply by willis102** [reply]
> The ability to load MayaReference from a proxy shape (create > USD) is something we are actively working on.

@kxl-adsk I know it's only been a few months since your previous response, but I'm wondering if there is any more news that can be shared externally on this yet, especially in regards to timeline.  Thanks!

**Reply by kxl-adsk** [reply]
Hi @willis102. I no longer work at Autodesk. I hope someone from usd team can comment. (FYI. @neilh-adsk, @seando-adsk)

Cheers,
KxL

**Reply by neilh-adsk** [reply]
Hi @willis102 - I'm not sure I followed the whole thread, but bringing a Maya reference and then caching that working to USD is something that is available now.  It will be officially turned on in the "Released" build with the next version.  Is this what you are after - Maya Reference to USD round trips etc ?
